### PR TITLE
DM-35391: Improve query-data-ids command

### PR DESCRIPTION
**Pull request by timj**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/script/queryDataIds.py [hunk 1]
```diff
@@ -78,6 +82,26 @@ def queryDataIds(repo, dimensions, datasets, where, collections, order_by, limit
     # Docstring for supported parameters is the same as Registry.queryDataIds
 
     butler = Butler(repo)
+
+    if not dimensions and datasets and collections:
```
Comment (parejkoj) [inline python/lsst/daf/butler/script/queryDataIds.py]: I'd rather have parentheses here: I never get the ordering of not/or/and.

File: python/lsst/daf/butler/script/queryDataIds.py [hunk 2]
```diff
@@ -89,8 +113,8 @@ def queryDataIds(repo, dimensions, datasets, where, collections, order_by, limit
...
         results.limit(limit, offset)
 
-    if len(results.graph) > 0:
+    if results.count() > 0 and len(results.graph) > 0:
         table = _Table(results)
-        return table.getAstropyTable(not order_by)
+        return table.getAstropyTable(not order_by), None
     else:
-        return None
+        return None, "\n".join(results.explain_no_results())
```
Comment (parejkoj) [inline python/lsst/daf/butler/script/queryDataIds.py:128]: Is this tuple the convention for returning no results explanations elsewhere in the butler? I feel like raising `NoResults` (or some other custom exception) with the explanation in the payload would be more pythonic. Looking at the tests, it certainly would simplify this interface.

File: tests/test_cliCmdQueryDataIds.py [hunk 3]
```diff
@@ -63,28 +63,31 @@ def tearDown(self):
...
+        self.assertFalse(msg)
         self.assertAstropyTablesEqual(res, expected)
 
     def testNull(self):
         "Test asking for nothing."
-        res = self._queryDataIds(self.root)
-        self.assertEqual(res, None)
+        res, msg = self._queryDataIds(self.root)
+        self.assertEqual(res, None, msg)
+        self.assertFalse(msg)
```
Comment (parejkoj) [inline tests/test_cliCmdQueryDataIds.py]: `msg` here is `None`, so use `assertIsNone(msg)`. (and more below)

File: tests/test_cliCmdQueryDataIds.py [hunk 4]
```diff
@@ -63,28 +63,31 @@ def tearDown(self):
 
     def testDimensions(self):
         """Test getting a dimension."""
-        res = self._queryDataIds(self.root, dimensions=("visit",))
+        res, msg = self._queryDataIds(self.root, dimensions=("visit",))
         expected = AstropyTable(
             array((("R", "DummyCamComp", "d-r", 423), ("R", "DummyCamComp", "d-r", 424))),
             names=("band", "instrument", "physical_filter", "visit"),
         )
+        self.assertFalse(msg)
```
Comment (parejkoj) [inline tests/test_cliCmdQueryDataIds.py:71]: `msg` here is `None`, so use `assertIsNone(msg)`.

File: python/lsst/daf/butler/script/queryDataIds.py [hunk 5]
```diff
@@ -78,6 +82,26 @@ def queryDataIds(repo, dimensions, datasets, where, collections, order_by, limit
...
+        graph = None
+        dataset_types = list(butler.registry.queryDatasetTypes(datasets))
+        for dataset_type in dataset_types:
+            if graph is None:
+                graph = dataset_type.dimensions
+            else:
+                graph = graph & dataset_type.dimensions
+            _LOG.debug("Dimensions now %s from %s", set(graph.names), dataset_type.name)
+            if not graph:
+                break
```
Comment (parejkoj) [inline python/lsst/daf/butler/script/queryDataIds.py:105]: It took me a bit to work out the logic here (using the other `if not graph` below: maybe an explanatory comment just above? `graph = graph & dimensions` did not obviously parse in my head as something that could make `graph` an empty set (it's a set, right?).

File: python/lsst/daf/butler/script/queryDataIds.py [hunk 6]
```diff
@@ -78,6 +82,26 @@ def queryDataIds(repo, dimensions, datasets, where, collections, order_by, limit
     # Docstring for supported parameters is the same as Registry.queryDataIds
 
     butler = Butler(repo)
+
+    if not dimensions and datasets and collections:
+        # Determine the dimensions relevant to all given dataset types.
+        graph = None
+        dataset_types = list(butler.registry.queryDatasetTypes(datasets))
+        for dataset_type in dataset_types:
+            if graph is None:
```
Comment (parejkoj) [inline python/lsst/daf/butler/script/queryDataIds.py:93]: `graph` here is a `set`, no? Maybe assign it to an empty set above, as that would make the logic of this loop more obvious.

File: python/lsst/daf/butler/script/queryDataIds.py [hunk 7]
```diff
@@ -78,6 +82,26 @@ def queryDataIds(repo, dimensions, datasets, where, collections, order_by, limit
...
+                graph = dataset_type.dimensions
+            else:
+                graph = graph & dataset_type.dimensions
+            _LOG.debug("Dimensions now %s from %s", set(graph.names), dataset_type.name)
+            if not graph:
+                break
+
+        if not graph:
+            names = [d.name for d in dataset_types]
+            return None, f"No dimensions in common for specified dataset types ({names})"
```
Comment (parejkoj) [inline python/lsst/daf/butler/script/queryDataIds.py:109]: As I comment below on the other `return` statements, I think this makes more sense raising a custom exception or `NoResults`.

**Review by parejkoj — CHANGES_REQUESTED** [review]
Some suggestions on clarifying the logic, and I'd rather it raised an exception than returned a tuple that always had one element be `None`.

File: python/lsst/daf/butler/script/queryDataIds.py (hunk 1 above)
Comment (timj) [reply python/lsst/daf/butler/script/queryDataIds.py]: `black` controls the formatting and decides whether parentheses are required or not. I can move the `not dimensions` to the end though.

File: python/lsst/daf/butler/script/queryDataIds.py (hunk 2 above)
Comment (timj) [reply python/lsst/daf/butler/script/queryDataIds.py:128]: This is not a butler convention -- it's trying to bridge the click to script implementation boundary. A specialist exception for this one function seems overkill to me.

File: python/lsst/daf/butler/script/queryDataIds.py (hunk 5 above)
Comment (timj) [reply python/lsst/daf/butler/script/queryDataIds.py:105]: Okay. This line is a short circuit given that once you have an empty set the logical AND can never result in content.

File: python/lsst/daf/butler/script/queryDataIds.py (hunk 6 above)
Comment (timj) [reply python/lsst/daf/butler/script/queryDataIds.py:93]: If I start with an empty set then ANDing anything will result in an empty set. I use `None` to indicate that nothing has been looked at yet. The alternative would be to build a default set with all possible dimensions in it. `None` seemed easier.

**Review by parejkoj — APPROVED** [review]
Thanks, those added comments help a lot with understanding the logic.

I'd still prefer a custom exception to a "one is always None" return tuple, but this works.

**Pull request merged**